packages/core/components/DataSourcePrompt/index.tsx

```diff
@@ -33,53 +32,49 @@ export default function DataSourcePrompt(props: Props) {
     const dataSourceInfo = useSelector(interaction.selectors.getDataSourceInfoForVisibleModal);
     const { source: sourceToReplace, query } = dataSourceInfo || ({} as DataSourcePromptInfo);
 
-    const [dataSourceURL, setDataSourceURL] = React.useState("");
+    const [dataSource, setDataSource] = React.useState<Source>();
+    const [metadataSource, setMetadataSource] = React.useState<Source>();
+    const [hasMetadataSource, setHasMetadataSource] = React.useState(false);
     const [isDataSourceDetailExpanded, setIsDataSourceDetailExpanded] = React.useState(false);
 
-    const addOrReplaceQuery = (source: Source) => {
-        if (sourceToReplace) {
-            dispatch(selection.actions.replaceDataSource(source));
-        } else if (query) {
-            dispatch(selection.actions.changeDataSources([...selectedDataSources, source]));
-        } else {
-            dispatch(
-                selection.actions.addQuery({
-                    name: `New ${source.name} Query`,
-                    parts: { sources: [source] },
-                })
-            );
+    const onSubmit = (dataSource?: Source, metadataSource?: Source) => {
+        if (dataSource && (metadataSource || !hasMetadataSource)) {
+            if (sourceToReplace) {
+                dispatch(selection.actions.replaceDataSource(dataSource));
+                dispatch(selection.actions.changeSourceMetadata(metadataSource));
+            } else if (query) {
+                dispatch(selection.actions.changeDataSources([...selectedDataSources, dataSource]));
+                dispatch(selection.actions.changeSourceMetadata(metadataSource));
```

Comment:
I'm seeing an intermittent bug where sometimes if I go to add a new query with a new datasource, instead of creating a new query it just replaces?removes? the metadata source from the current query. Can't always reproduce it though

Reply:
I wasn't able to reproduce this -- lmk if you keep seeing it though